### PR TITLE
Feature/skip decorator

### PR DESCRIPTION
**Pull request by mateimicu**

### CODE REVIEW

File: argus/exceptions.py [hunk 1]
```diff
@@ -38,3 +38,8 @@ class ArgusHeatTeardown(ArgusError):
 class ArgusEnvironmentError(ArgusError):
     """Base class for errors related to the argus environment."""
     pass
+
+
+class ArgusLogicError(ArgusError):
```
Comment (ader1990) [inline argus/exceptions.py]: name the error as per the description: ArgusInvalidState

Comment (ader1990) [reply argus/exceptions.py]: although I'd put here something like; ArgusInvalidDecorator

File: argus/util.py [hunk 2]
```diff
@@ -24,11 +24,16 @@
 import subprocess
 import sys
 import time
+import unittest
 
 import six
 
+from argus import log as argus_log
+from argus.tests import base as base_testcase
```
Comment (ader1990) [inline argus/util.py]: I would not add a unittest dependecy in the argus utils, as this module should be independent to other modules used in argus

File: argus/util.py [hunk 3]
```diff
@@ -362,3 +367,57 @@ def get_command(command, command_type=None):
...
+                target_self = kwargs.get("self", args[0])
+            except IndexError:
+                raise exceptions.ArgusLogicError(
+                    "Incompatible use of decorator %s on %s",
+                    "skip_on_os", function.__name__)
+
+            instance_os_type = None
+            if isinstance(target_self, base_testcase.BaseTestCase):
+                try:
+                    instance_os_type = get_os_type(target_self)
```
Comment (ader1990) [inline argus/util.py]: here we can use duck typing -- if a class has fly method, we can safely say is a flying bird. this way, we can omit the isinstance call and remove the basetestcase dependency

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/util.py [hunk 4]
```diff
@@ -362,3 +367,57 @@ def get_command(command, command_type=None):
...
+        """
+        self._target_os_list = target_os_list
+        self._reason = reason
+
+    def __call__(self, function):
+        """Return the `function` decorated."""
+        @six.wraps(function)
+        def wrapper(*args, **kwargs):
+            try:
+                target_self = kwargs.get("self", args[0])
```
Comment (alexcoman) [inline argus/util.py]: I don't think that somebody will pass `self=self`. Taking that into consideration we can safely asume that first positional argument is a reference to the instance.

File: argus/util.py [hunk 5]
```diff
@@ -362,3 +367,57 @@ def get_command(command, command_type=None):
...
+            instance_os_type = None
+            if isinstance(target_self, base_testcase.BaseTestCase):
+                try:
+                    instance_os_type = get_os_type(target_self)
+                except AttributeError as ex:
+                    raise exceptions.ArgusLogicError(
+                        ("The OS type has not been determined yet in %s,"
+                         " error message $s"), function.__name__, ex)
+            else:
+                raise exceptions.ArgusLogicError(
```
Comment (alexcoman) [inline argus/util.py]: I think a warning will be enough. We can return the  wrapped method.

**Review by alexcoman — CHANGES_REQUESTED** [review]

File: argus/exceptions.py [hunk 6]
```diff
@@ -38,3 +38,8 @@ class ArgusHeatTeardown(ArgusError):
 class ArgusEnvironmentError(ArgusError):
     """Base class for errors related to the argus environment."""
     pass
+
+
+class ArgusInvalidDecorator(ArgusError):
```
Comment (ader1990) [inline argus/exceptions.py]: Add an Error suffix to the class name

File: argus/exceptions.py [hunk 7]
```diff
@@ -38,3 +38,8 @@ class ArgusHeatTeardown(ArgusError):
...
+    """Error related to invalid states in the Argus-CI code."""
```
Comment (ader1990) [inline argus/exceptions.py]: add a corresponding docstring to the error

File: argus/util.py [hunk 8]
```diff
@@ -362,3 +366,57 @@ def get_command(command, command_type=None):
         True: WINDOWS_NANO
     }
 }
+
+
+def get_os_type(instance):
+    """Get the OS type from a `Recipe` or `TestCase` instance."""
+    # pylint: disable=protected-access
+    return instance._backend.remote_client.manager.os_type
```
Comment (ader1990) [inline argus/util.py]: is this the only way to get the OS type? It would be nice to be a public property/getter method

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/util.py [hunk 9]
```diff
@@ -362,3 +366,51 @@ def get_command(command, command_type=None):
...
 }
+
+
+class skip_on_os(object):
+    """Decorator to skip a method on a specific OS."""
+
+    def __init__(self, target_os_list, reason=""):
+        """Skip the method decorated if we are using the `os_type`.
+
+        :param target_os_listos_type: A list with OS version in on which
```
Comment (ader1990) [inline argus/util.py]: target_os_list*
in on -> on

File: argus/exceptions.py [hunk 10]
```diff
@@ -38,3 +38,8 @@ class ArgusHeatTeardown(ArgusError):
 class ArgusEnvironmentError(ArgusError):
     """Base class for errors related to the argus environment."""
     pass
+
+
+class ErrorArgusInvalidDecorator(ArgusError):
+    """The `skip_on_os` decorator was used improperly."""
```
Comment (ader1990) [inline argus/exceptions.py]: Exception triggered when a decorator has been improperly used.

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/util.py [hunk 11]
```diff
@@ -362,3 +366,51 @@ def get_command(command, command_type=None):
         True: WINDOWS_NANO
     }
 }
+
+
+class skip_on_os(object):
+    """Decorator to skip a method on a specific OS."""
+
+    def __init__(self, target_os_list, reason=""):
```
Comment (stefan-caraiman) [inline argus/util.py:374]: a default of `Unknown` seems more appropriate than an empty string.

File: argus/util.py [hunk 12]
```diff
@@ -362,3 +366,51 @@ def get_command(command, command_type=None):
...
+        @six.wraps(function)
+        def wrapper(*args, **kwargs):
+            try:
+                target_self = args[0]
+            except IndexError:
+                raise exceptions.ErrorArgusInvalidDecorator(
+                    "Incompatible use of decorator %s on %s",
+                    "skip_on_os", function.__name__)
+
+            instance_os_type = None
```
Comment (stefan-caraiman) [inline argus/util.py]: why initialize with `instance_os_type` since the function will return if used incorrectly?

File: argus/exceptions.py [hunk 13]
```diff
@@ -38,3 +38,8 @@ class ArgusHeatTeardown(ArgusError):
 class ArgusEnvironmentError(ArgusError):
     """Base class for errors related to the argus environment."""
     pass
+
+
+class ErrorArgusInvalidDecorator(ArgusError):
```
Comment (stefan-caraiman) [inline argus/exceptions.py]: shouldn't we keep the same naming convention for argus errors? something like `ArgusInvalidDecoratorError` seems more fit.

File: argus/util.py (hunk 11 above)
Comment (ader1990) [reply argus/util.py:374]: it s okay like this

**Review by ader1990 — CHANGES_REQUESTED** [review]

**Review by ader1990 — APPROVED** [review]

**Pull request merged**